processor/src/main/java/org/mapstruct/ap/internal/model/CollectionAssignmentBuilder.java

```diff
@@ -0,0 +1,163 @@
+/**
+ *  Copyright 2012-2017 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
```

Comment:
I was thinking of using a chain of responsibility. Each chain element is a current wrapper and triggering on its own criteria (the current if-then-elses).